### ISSUE TITLE
Unable to push publish plugins

### ISSUE THREAD

**Issue by swicken-dotcms** [dotCMS/core#21568]
**Describe the bug**
Attempting to push publish plugins results in an error attempting to generate a directory

```
[20/01/22 14:43:49:427 UTC]  INFO publishing.PublisherAPIImpl: Start of Bundler: OSGIBundler
[20/01/22 14:43:49:445 UTC] ERROR publishing.PublisherAPIImpl: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
com.dotcms.publishing.DotBundleException: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:119) ~[ee-22.01.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:118) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl$AjcClosure1.run(PublisherAPIImpl.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:47) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute_aroundBody0(PublisherQueueJob.java:186) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob$AjcClosure1.run(PublisherQueueJob.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute(PublisherQueueJob.java:112) ~[dotcms_22.01_999999.jar:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:223) ~[dot.quartz-all-1.8.6_2.jar:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549) ~[dot.quartz-all-1.8.6_2.jar:?]
Caused by: java.io.IOException: Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFile(FileUtils.java:777) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:706) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:102) ~[ee-22.01.jar:?]
	... 17 more
[20/01/22 14:43:49:445 UTC] ERROR business.PublisherQueueJob: Unable to publish Bundle '01FSVZ34DPME6FZT8ZEE90KSXC': com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
com.dotcms.publishing.DotPublishingException: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:154) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl$AjcClosure1.run(PublisherAPIImpl.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:47) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute_aroundBody0(PublisherQueueJob.java:186) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob$AjcClosure1.run(PublisherQueueJob.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute(PublisherQueueJob.java:112) ~[dotcms_22.01_999999.jar:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:223) ~[dot.quartz-all-1.8.6_2.jar:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549) ~[dot.quartz-all-1.8.6_2.jar:?]
Caused by: com.dotcms.publishing.DotBundleException: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:119) ~[ee-22.01.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:118) ~[dotcms_22.01_999999.jar:?]
	... 16 more
Caused by: java.io.IOException: Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFile(FileUtils.java:777) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:706) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:102) ~[ee-22.01.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:118) ~[dotcms_22.01_999999.jar:?]
	... 16 more
[20/01/22 14:43:49:445 UTC]  INFO util.PushPublishLogger: class com.dotcms.publisher.business.PublisherQueueJob : Status Update: Failed to bundle '01FSVZ34DPME6FZT8ZEE90KSXC'
```

Reproduced on: 21.06.6, 22.01
Related Ticket: https://dotcms.zendesk.com/agent/tickets/106313

**To Reproduce**
1. Create a sender/receiver environment
2. Add a plugin to the sender
3. Attempt to push publish that plugin

**Expected behavior**
The plugin should push successfully.

discussion https://github.com/dotCMS/core/discussions/21590

**Comment by wezell** [comment]
This does not work in docker environments.  When bundling OSGI plugins, we should use the normal `/asset/bundle` folder as that folder is shared and gets automatically cleaned up etc...

**Comment by bryanboza** [comment]
Hey guys I have a couple of questions here that maybe we need to take in consideration... 

Does this work with the OSGI, static ones or both? 

Also, what should expect in case you already have the plugin in the receiver side, should we have integrity checker for plugins? or just should replace the existing one without confirmation ?

**Comment by swicken-dotcms** [comment]
Additional ticket: https://dotcms.zendesk.com/agent/tickets/106537

**Comment by victoralfaro-dotcms** [comment]
PRs:
https://github.com/dotCMS/tomcat/pull/22
https://github.com/dotCMS/enterprise/pull/854
https://github.com/dotCMS/core/pull/21752

**Comment by erickgonzalez** [comment]
Need to test in 5.3.8 series, probably not needed since was introduced with the changes in PP (21.05ish)

**Comment by dsilvam** [comment]
PR ee; https://github.com/dotCMS/enterprise/pull/866

**Comment by dsilvam** [comment]
PR ee: https://github.com/dotCMS/enterprise/pull/866

**Comment by freddyucv** [comment]
**Pass Internal QA**

it was test with this plugins: https://github.com/dotCMS/plugin-seeds/tree/master/OSGi/com.dotcms.rest

- Start the sender in 8080 port.
- Start receiver in 8090 port.
- Upload the bundle in the sender.
- Test the Plugins with: curl -XPUT http://localhost:8080/api/example
   Response: {"result":"anonymous user anonymous PUT!"}
- Test the Plugins with: curl --head -XPUT http://localhost:8090/api/example
   Response: 404 HTTP Code
- PP the Plugins to the receiver
- Test Again the Plugins with: curl --head -XPUT http://localhost:8090/api/example
   Response: {"result":"anonymous user anonymous PUT!"}
- Change the code here: 

https://github.com/dotCMS/plugin-seeds/blob/master/OSGi/com.dotcms.rest/src/main/java/com/dotcms/plugin/rest/ExampleResource.java#L36

Change " PUT!\" by " PUT! With changes"
- Upload the bundle again in the sender and PP it to the receiver
- Now test with: curl -XPUT http://localhost:8090/api/example
Response: {"result":"anonymous user anonymous PUT! With changes"}

**Comment by freddyucv** [comment]
> Hey guys I have a couple of questions here that maybe we need to take in consideration...
> 
> Does this work with the OSGI, static ones or both?
> 
> Also, what should expect in case you already have the plugin in the receiver side, should we have integrity checker for plugins? or just should replace the existing one without confirmation ?
> 
> Taking in consideration this change can broke some features in case of a wrong configuration, maybe at least we should create a new property to enable/disable permissions to override the configuration.

This is a bug is not a new issue

**Comment by bryanboza** [comment]
We need some work here, every time you push any plugin we are getting this error in the log: 
https://gist.github.com/bryanboza/b40712d8d129c0fc8518288b3493d7c9

**Comment by bryanboza** [comment]
Fixed, tested on release-22.06 and now we don't have the error in the log

**Comment by swicken-dotcms** [comment]
Will not be backporting to LTS releases due to changes to bundle manifest files and an upgrade task requirement
